frontend/z3_axioms.py

```diff
@@ -0,0 +1,236 @@
+            And(subtype(left, Num), subtype(right, Num), subtype(result, Num))
+        ),
+
+        Implies(
+            And(subtype(left, Num), subtype(right, Num), stronger_num(left, right)),
+            result == left
+        ),
+
+        Implies(
+            And(subtype(left, Num), subtype(right, Num), stronger_num(right, left)),
+            result == right
+        )
+    ]
+
+
+def div(left, right, result):
+    return [
+        And(subtype(left, Num), subtype(right, Num)),
+        Implies(Or(left == Complex, right == Complex), result == Complex),
+        Implies(Not(Or(left == Complex, right == Complex)), result == Float)
+    ]
+
+
+def arithmetic(left, right, result):
+    return [
+        And(subtype(left, Num), subtype(right, Num)),
+        Implies(stronger_num(left, right), result == left),
+        Implies(stronger_num(right, left), result == right)
+    ]
+
+
+def bitwise(left, right, result):
+    return arithmetic(left, right, result) + [And(subtype(left, Int), subtype(right, Int))]
+
+
+def unary_invert(unary):
+    return [
+        subtype(unary, Int)
+    ]
+
+
+def unary_other(unary, result):
+    return [
+        subtype(unary, Num),
+        Implies(unary == Bool, result == Int),
+        Implies(unary != Bool, result == unary)
+    ]
+
+
+def if_expr(a, b, result):
+    return [
+        subtype(a, result),
+        subtype(b, result)
+    ]
+
+
+def index(indexed, index, result):
+    t = []
+    quantifiers_consts = [Const("tuples_q_{}".format(x), type_sort) for x in range(len(Tuples) - 1)]
+    for cur_len in range(1, len(Tuples)):
+        quants = quantifiers_consts[:cur_len]
+        t.append(Exists(quants, And(
+            indexed == Tuples[cur_len](*quants),
+            *[subtype(x, result) for x in quants]
+        ), patterns=[Tuples[cur_len](*quants)]))
+
+    return [
+        Or(
+            [indexed == Dict(index, result),
+             And(subtype(index, Int), indexed == List(result)),
+             And(subtype(index, Int), indexed == String, result == String),
+             And(subtype(index, Int), indexed == Bytes, result == Bytes)]
+            + t
+        )
+    ]
+
+
+def slice(lower, upper, step, sliced, result):
+    return [
+        And(subtype(lower, Int), subtype(upper, Int), subtype(step, Int), subtype(sliced, Seq), result == sliced)
+    ]
+
+
+def generator(iter, target):
+    # TODO tuples
+    return [
+        Or(
+            iter == List(target),
+            iter == Set(target),
+            And(iter == String, target == String),
+            And(iter == Bytes, target == Bytes),
+            Exists(x, iter == Dict(target, x), patterns=[Dict(target, x)])
+        )
+    ]
+
+
+def assignment(target, value):
+    return [
+        Implies(subtype(target, Num), And(subtype(value, Num), stronger_num(target, value))),
+        Implies(Not(subtype(target, Num)), subtype(value, target))
+    ]
+
+
+def assignment_target(target, value, position):
```

Comment:
Added docstrings to all functions.

As for the quantifiers, this method is for multiple-assignment (renamed method to give more describing name). And when it gets to tuple assignment, like `a, b = 1, "string"`, each target on the left hand side is treated individually, and accordingly, in order to assert with different tuple lengths, existential quantifiers had to be used.